File: tests/bun/tests/reversed.ts
import { makeTetradicNumber, getLowestTetradicNumberIndex } from '../lib';

const makeTetradicNumbers = (num: number) => {
  const numbers = []
  let index = 0

  while (true) {
    const tetradicNumber = makeTetradicNumber(index);
    if (tetradicNumber > num) {
      break;
    }

    numbers.push(tetradicNumber);
    index++
  }

  return numbers;
}
function findSums(inputNumber: number): number[] | null {
  const tetradicNumbers = makeTetradicNumbers(inputNumber);
  const length = tetradicNumbers.length;

  for (let i1 = length - 1; i1 >= 0; i1--) {
    const n1 = tetradicNumbers[i1];
    if (n1 === inputNumber) {
      return [n1];
    }

    if (n1 > inputNumber) {
      continue;
    }

    for (let i2 = length - 1; i2 >= i1; i2--) {
      const n2 = tetradicNumbers[i2];
      const required2 = inputNumber - n1 - n2;
      if (required2 === 0) {
        return [n1, n2];
      }

      if (required2 < 0) {
        continue;
      }

      for (let i3 = i2; i3 >= i1; i3--) {
        const n3 = tetradicNumbers[i3];
        const required3 = required2 - n3;
        if (required3 === 0) {
          return [n1, n2, n3];
        }

        if (required3 < 0) {
          continue;
        }

        for (let i4 = i3; i4 >= i1; i4--) {
          const n4 = tetradicNumbers[i4];
          const required4 = required3 - n4;
          if (required4 === 0) {
            return [n1, n2, n3, n4];
          }

          if (required4 < 0) {
            continue;
          }

          for (let i5 = i4; i5 >= i1; i5--) {
            const n5 = tetradicNumbers[i5];
            const required5 = required4 - n5;
            if (required5 === 0) {
              return [n1, n2, n3, n4, n5];
            }
          }
        }
      }
    }
  }

  return null;
}

function test(start: number, end: number, step: number) {
  for (let i = start; i < end; i++) {
    const result = findSums(i);

    if (result === null || result.reduce((acc, curr) => acc + curr, 0) !== i) {
      throw new Error(`No result for ${i}`);
    }
  }
}

const args = Bun.argv.slice(2).map(Number) as [number, number, number];

test(...args);

File: tests/python/tests/reversed.py
import sys

def make_tetradic_number(index):
  # Placeholder function, replace with actual implementation
  return (index * (index + 1) * (index + 2)) // 6

def get_lowest_tetradic_number_index(num):
  # Placeholder function, replace with actual implementation
  index = 0
  while make_tetradic_number(index) <= num:
    index += 1
  return index - 1

def make_tetradic_numbers(num):
  numbers = []
  index = 0

  while True:
    tetradic_number = make_tetradic_number(index)
    if tetradic_number > num:
      break

    numbers.append(tetradic_number)
    index += 1

  return numbers

def find_sums(input_number):
  tetradic_numbers = make_tetradic_numbers(input_number)
  length = len(tetradic_numbers)

  for i1 in range(length - 1, -1, -1):
    n1 = tetradic_numbers[i1]
    if n1 == input_number:
      return [n1]

    if n1 > input_number:
      continue

    for i2 in range(length - 1, i1 - 1, -1):
      n2 = tetradic_numbers[i2]
      required2 = input_number - n1 - n2
      if required2 == 0:
        return [n1, n2]

      if required2 < 0:
        continue

      for i3 in range(i2, i1 - 1, -1):
        n3 = tetradic_numbers[i3]
        required3 = required2 - n3
        if required3 == 0:
          return [n1, n2, n3]

        if required3 < 0:
          continue

        for i4 in range(i3, i1 - 1, -1):
          n4 = tetradic_numbers[i4]
          required4 = required3 - n4
          if required4 == 0:
            return [n1, n2, n3, n4]

          if required4 < 0:
            continue

          for i5 in range(i4, i1 - 1, -1):
            n5 = tetradic_numbers[i5]
            required5 = required4 - n5
            if required5 == 0:
              return [n1, n2, n3, n4, n5]

  return None

def test(start, end, step):
  for i in range(start, end):
    result = find_sums(i)

    if result is None or sum(result) != i:
      raise ValueError(f"No result for {i}")

if __name__ == "__main__":
  args = list(map(int, sys.argv[1:4]))
  test(*args)